vic/drivers/classic/src/vic_populate_model_state.c

```diff
@@ -0,0 +1,101 @@
+ * contents are checked to see if it agrees with the current simulation set-up,
+ * if so it is used to initialize the model state.  If no state file is
+ * provided the model initializes all variables with defaults and the user
+ * should expect to throw out the beginning of the simulation period as model
+ * start-up.
+ *
+ * @section LICENSE
+ *
+ * The Variable Infiltration Capacity (VIC) macroscale hydrological model
+ * Copyright (C) 2014 The Land Surface Hydrology Group, Department of Civil
+ * and Environmental Engineering, University of Washington.
+ *
+ * The VIC model is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *****************************************************************************/
+
+#include <vic_driver_classic.h>
+
+/******************************************************************************
+ * @brief    Initialize the model state (energy balance, water balance, and
+ *           snow components).
+ *****************************************************************************/
+void
+vic_populate_model_state(all_vars_struct *all_vars,
+                         filep_struct     filep,
+                         size_t           cellnum,
+                         soil_con_struct *soil_con,
+                         veg_con_struct  *veg_con,
+                         lake_con_struct  lake_con)
+{
+    extern option_struct options;
+
+    size_t               Nveg;
+    int                  tmp_lake_idx;
+
+    cell_data_struct   **cell;
+    energy_bal_struct  **energy;
+    lake_var_struct     *lake;
+    snow_data_struct   **snow;
+    veg_var_struct     **veg_var;
+
+    cell = all_vars->cell;
+    energy = all_vars->energy;
+    lake = &all_vars->lake_var;
+    snow = all_vars->snow;
+    veg_var = all_vars->veg_var;
+
+    Nveg = veg_con[0].vegetat_type_num;
+
+    // Initialize all data structures to 0
+    initialize_soil(cell, soil_con, Nveg);
+    initialize_snow(snow, Nveg);
+    initialize_veg(veg_var, Nveg);
+    if (options.LAKES) {
+        tmp_lake_idx = lake_con.lake_idx;
+        if (tmp_lake_idx < 0) {
+            tmp_lake_idx = 0;
+        }
+        initialize_lake(lake, lake_con, soil_con, &(cell[tmp_lake_idx][0]),
+                        false);
+    }
+    initialize_energy(energy, Nveg);
+
```

Comment:
Currently, in the following, `if (options.INIT_STATE)` reads in initial state file and modifies some direct states (i.e., states that cannot be derived and must be saved), supposedly; `else` if no initial state, then a part of states are modified. Then `compute_derived_state_vars` again modifies a part of states that are derived. Then the rest of the states might be derived and modified in the vic_run core code.

To me this seems a bit confusing because it's not very clear which state variables are 'direct' and must be saved and read from initial state file, and which are 'derived'; and it is possible that some 'direct' states are not correctly saved and read, causing inexact restart (as I'm seeing now for a simple bare soil case). Would it be possible to separate 'direct' and 'derived' states in a cleaner way? For example, maybe not call `initialize_*` functions at all, and then in `if (options.INIT_STATE)`, read in all 'direct' states from state file (as it is doing now supposedly); in `else`, set the same set of 'direct' states to their default values (I guess currently many of the default values are zero, so they are assigned above in `initialize_*`). In this way, all the 'derived' variables will either be assigned in `compute_derived_state_vars`, or in the core `vic_run` computation; but if any 'derived' variables are used before being assigned, then there would be an error and we should avoid it from happening.

These are just some thoughts based on my understanding of the code structure, and not sure if makes sense at all...